### PR TITLE
feat: add take grad

### PR DESCRIPTION
**Pull request by polvalente**
Similar to #460

### CODE REVIEW

File: nx/test/nx/defn/grad_test.exs [hunk 1]
```diff
@@ -2503,6 +2503,141 @@ defmodule Nx.Defn.GradTest do
...
+                 Nx.tensor([
+                   [
+                     [[0], [1], [2]],
+                     [[2], [1], [0]],
+                     [[0], [1], [2]]
+                   ]
+                 ])
+               )
+
+      assert Nx.tensor([
```
Comment (polvalente) [inline nx/test/nx/defn/grad_test.exs:2618]: This is mostly working, but this last assertion fails

File: nx/lib/nx/defn/grad.ex [hunk 2]
```diff
@@ -507,6 +507,62 @@ defmodule Nx.Defn.Grad do
...
+  defp grad(:take, [t, i, axis], _ans, g, cache) do
+    axes_range = 0..(tuple_size(t.shape) - 1)//1
+
+    indices_shape =
+      axes_range
+      |> Enum.map(fn
+        ^axis -> Tuple.to_list(i.shape)
+        _ -> 1
+      end)
+      |> List.flatten()
```
Comment (josevalim) [inline nx/lib/nx/defn/grad.ex:519]:
```suggestion
      |> Enum.flat_map(fn
        ^axis -> Tuple.to_list(i.shape)
        _ -> [1]
      end)
```

A similar technique can be employed below if you want.

File: nx/lib/nx/defn/grad.ex [hunk 3]
```diff
@@ -507,6 +507,62 @@ defmodule Nx.Defn.Grad do
     to_grad(t, g, cache)
   end
 
+  defp grad(:take, [t, i, axis], _ans, g, cache) do
+    axes_range = 0..(tuple_size(t.shape) - 1)//1
```
Comment (josevalim) [inline nx/lib/nx/defn/grad.ex:511]:
```suggestion
    axes_range = 0..(tuple_size(t.shape) - 1)//1
```

Do we use `Nx.rank(t)` or `tuple_size(t.shape)` in this file? make sure we are following one "standard" or the other. :)

**Review by josevalim — APPROVED** [review]
Looking gorgeous, just some small comments added!

File: nx/lib/nx/defn/grad.ex (hunk 3 above)
Comment (polvalente) [reply nx/lib/nx/defn/grad.ex:511]: Nx.rank :)

I'll change this and possibly other occurences

**Review by josevalim — APPROVED** [review]

**Pull request merged**